include/swift/Frontend/Frontend.h

```diff
@@ -107,6 +107,12 @@ class CompilerInvocation {
 
   CodeCompletionCallbacksFactory *CodeCompletionFactory = nullptr;
 
+  /// This \c CancellationFlag will be set on the \c ASTContext constructed from
+  /// the invocation. If the shared pointer is not null and is set to \c true,
+  /// all operations operating on that \c ASTContext are requested to cancel at
+  /// the next possible opportunity.
+  std::shared_ptr<std::atomic<bool>> CancellationFlag = nullptr;
```

Comment:
What's the motivation of setting this in `CompilerInvocation`? Especially in fast-completion scenario, I feel a little uncomfortable that `CompilerInstance` keep holding the `CancellationFlag` of the _original_ request.

In the fast completion code path, you manually do
```cpp
CI.getASTContext().CancellationFlag = CancellationFlag;
```
I think this is OK, and other ASTManager requests can do this too after constructing ASTContext.

Related: I'm not sure we really need to change `ASTContext::get()`. Other parameters are required because they are references. But `CancellationFlag` can be initialized with `nullptr`.

Reply:
Two very excellent questions. I think I set `CancellationFlag` on `CompilerInvocation` because that’s what we used to do when it was part of the type checker options. But setting it after creating the compiler instance makes way more sense.

Also: There really is no need to change `ASTContext::get()` I think I wanted to make sure I didn’t miss to set a cancellation flag where I should have set one, but initializing it to `nullptr` and setting it afterwards is much cleaner.